pkg/signup/keymanager.go

```diff
@@ -0,0 +1,172 @@
+package signup
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"github.com/codeready-toolchain/registration-service/pkg/configuration"
+	"gopkg.in/square/go-jose.v2"
+)
+
+// PublicKey represents an RSA public key with a Key ID
+type PublicKey struct {
+	KeyID string
+	Key   *rsa.PublicKey
+}
+
+// JSONKeys the remote keys encoded in a json document
+type JSONKeys struct {
+	Keys []interface{} `json:"keys"`
+}
+
+// KeyManager manages the public keys for token validation.
+type KeyManager struct {
+	config *configuration.Registry
+	logger *log.Logger
+	keyMap map[string]*rsa.PublicKey
+}
+
+// NewKeyManager creates a new KeyManager and retrieves the public keys from the given URL.
+func NewKeyManager(logger *log.Logger, config *configuration.Registry) (*KeyManager, error) {
+	if logger == nil {
+		return nil, errors.New("no logger given when creating KeyManager")
+	}
+	if config == nil {
+		return nil, errors.New("no config given when creating KeyManager")
+	}
+	isTesting := config.IsTestingMode()
+	keysEndpointURL := config.GetAuthClientPublicKeysURL()
+	km := &KeyManager{
+		logger: logger,
+		config: config,
+		keyMap: make(map[string]*rsa.PublicKey),
+	}
+	// fetch raw keys
+	var keys []*PublicKey
+	var err error
+	if !isTesting {
+		keys, err = km.fetchKeys(keysEndpointURL)	
+	} else {
+		keysRaw := config.GetViperInstance().Get("testkeys")
```

Comment:
what about adding a method to get these keys from the config, instead of having to use the underlying `viper` lib?

Reply:
Not in this case. This code is only used in the test run where there is no full config instance but an "injected" one. Also, we don't want config values that are only used for testing to be mandatory in a production config, so the "main" config should not be used to distribute the test keys.

Reply:
what about replacing the keys in the tests, instead of using a different set of keys?

Reply:
Not possible as those are retrieved from an http endpoint and this is the whitebox test. There is another issue here, there are intentionally no setter for config values. When we want to set a value from the test, we need to access the underlying infra.